implementation.js

```diff
@@ -28,6 +61,31 @@ module.exports = function from(arrayLike) {
 	var A = ES.IsCallable(C) ? ES.ToObject(new C(len)) : new Array(len);
 	var k = 0;
 	var kValue, mappedValue;
+
+	// variables for rebuilding array from iterator
+	var arrayFromIterable;
+
+	// if it is an iterable treat like one
+	arrayFromIterable = parseIterable(arrayLike);
+
+	//if it is a Map or a Set then handle them appropriately
+	if (
+		typeof arrayLike.entries === 'function'
+		&& typeof arrayLike.values === 'function'
+	) {
+		if (arrayLike.constructor.name === 'Set' && 'values' in Set.prototype) {
```

Comment:
I definitely don't like the idea of using `constructor`, which is highly unreliable.

I think we need a very robust way of detecting Maps and Sets which works in every engine, at every stage of `for..of` + `Symbol.iterator` support, that users aren't capable of faking.